src/onemod/pipeline.py

```diff
@@ -0,0 +1,144 @@
+       other stages (with the exception of any preprocessing stages)
+    * `data` is only required if using grouped stages
+    * TODO: Define dependencies between stages
+    * TODO: DAG creation, validation
+    * TODO: Run locally or on Jobmon
+
+    """
+
+    name: str
+    config: PipelineConfig
+    directory: Path
+    data: Path | None = None
+    groupby: set[str] = set()
+    _stages: dict[str, Stage] = {}  # set by Pipeline.add_stage
+
+    @computed_field
+    @property
+    def stages(self) -> set[str]:
+        return self._stages
+
+    def model_post_init(self, *args, **kwargs) -> None:
+        if not self.directory.exists():
+            self.directory.mkdir(parents=True)
+
+    @classmethod
+    def from_json(cls, filepath: Path | str) -> Pipeline:
+        """Load pipeline object from JSON file."""
+        with open(filepath, "r") as f:
+            pipeline_json = json.load(f)
+        stages = pipeline_json.pop("stages", None)
+        pipeline = cls(**pipeline_json)
+        if stages is not None:
+            pipeline.add_stages(stages)
+        return pipeline
+
+    def to_json(self, filepath: Path | str | None = None) -> None:
+        """Save pipeline object as JSON file."""
+        filepath = filepath or self.directory / (self.name + ".json")
+        with open(filepath, "w") as f:
+            f.write(self.model_dump_json(indent=4))
+
+    def add_stages(
+        self, stages: list[Stage | str], filepath: Path | str | None = None
+    ) -> None:
+        """Add stages to pipeline."""
+        for stage in stages:
+            self.add_stage(stage, filepath)
+
+    def add_stage(
+        self, stage: Stage | str, filepath: Path | str | None = None
+    ) -> None:
+        """Add stage to pipeline."""
+        if isinstance(stage, str):
+            stage = self._stage_from_json(stage, filepath)
+        self._add_stage(stage)
+
+    def _add_stage(self, stage: Stage) -> None:
+        """Add stage object to pipeline."""
+        if stage.name in self.stages:
+            raise ValueError(f"stage '{stage.name}' already exists")
+        stage.config.update(self.config)
+        stage.directory = self.directory / stage.name
+        if isinstance(stage, GroupedStage):
+            if self.data is None:
+                raise ValueError("data is required for GroupedStage")
+            stage.groupby.update(self.groupby)
+            stage.create_stage_subsets(self.data)
+        if isinstance(stage, CrossedStage):
+            if stage.config.crossable_params:
+                stage.create_stage_params()
+        self._stages[stage.name] = stage
+
+    def _stage_from_json(self, stage: str, filepath: Path | str) -> Stage:
+        """Load stage object from JSON."""
+        if stage in self.stages:
+            raise ValueError(f"stage '{stage}' already exists")
+        with open(filepath, "r") as f:
+            stage_json = json.load(f)
+        if stage_type := stage_json["type"] in STAGE_DICT:
```

Comment:
I might need some convincing on the use of `STAGE_DICT` here. I'm not exactly sure what a best alternative looks like at this moment, but for some reason I feel like within the core Pipeline component there might be a nicer alternative to relying on a static dictionary of stage types. Might be missing something though

Reply:
Agreed, removed for now; we can figure out an easier way to load built-in stages later